File: python/transpose/transpose.py
from itertools import zip_longest

def transpose(lines):
    """
    Transpose the input text, switching rows and columns while preserving characters.

    Args:
        lines (str): The input text containing multiple lines.

    Returns:
        str: The transposed text.

    Example:
        Input:
        "ABC\nDEF"

        Output:
        "AD\nBE\nCF"

    If the input has rows of different lengths, this function pads to the left with spaces
    to ensure that all characters from the input are present in the transposed output.

    """
    # Split the input text into rows
    rows = lines.split("\n")

    # Pad each line with spaces to match the length of the longest line
    lines = list(line.ljust(max(len(l) for l in rows[i+1:]), " ")
                 for i, line in enumerate(rows[:-1])) + [rows[-1]]

    # Transpose the lines and join them back into text
    transposed_text = "\n".join("".join(row) for row in
                                zip_longest(*lines, fillvalue=""))

    return transposed_text

File: javascript/transpose/transpose.js
export const transpose = (textstring) => {
  // Split the text into lines
  const lines = textstring.split("\n");

  // Find the length of the longest line
  const maxLength = lines.reduce((max, line) => Math.max(max, line.length), 0);

  // Pad all lines with spaces to make them the same length
  lines.forEach((line, i) => (lines[i] = line.padEnd(maxLength, " ")));

  // Transpose the text by swapping characters in each column
  let transposedText = "";
  for (let i = 0; i < maxLength; i++) {
    for (let line of lines) {
      transposedText += line[i];
    }
    transposedText += "\n";
  }

  // Format the transposed text as a string and return it
  return transposedText.trim();
};
